File: src/main/java/com/rtomyj/yugiohAPI/service/product/ProductService.java
package com.rtomyj.yugiohAPI.service.product;

import com.rtomyj.yugiohAPI.dao.database.Dao;
import com.rtomyj.yugiohAPI.helper.enumeration.products.ProductType;
import com.rtomyj.yugiohAPI.model.product.Product;
import com.rtomyj.yugiohAPI.model.product.Products;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Service;

@Service
public class ProductService
{

	@Autowired
	@Qualifier("jdbc")
	private Dao dao;


	public Products getProductsByLocale(final String locale)
	{

		final Products allProductsByLocale = new Products();
		allProductsByLocale.setProducts(dao.getProductsByLocale(locale));
		allProductsByLocale.setLinks();

		return allProductsByLocale;

	}

	public Products getProductsByLocaleAndProductType(final ProductType productType, final String locale)
	{

		final Products products = dao.getAllProductsByType(productType, locale);
		products.setLocale(locale);
		products.setProductType(productType);
		products.setLinks();

		return products;

	}


	public Product getProductByLocale(final String packId, final String locale)
	{

		final Product product = dao.getProductInfo(packId, locale);
		product.getProductContent().addAll(dao.getProductContents(packId, locale));
		product.setLinks();

		return product;

	}

}

File: src/main/kotlin/com/rtomyj/skc/find/product/ProductService.kt
package com.rtomyj.skc.find.product

import com.rtomyj.skc.browse.product.dao.ProductDao
import com.rtomyj.skc.model.MonsterAssociation
import com.rtomyj.skc.model.Product
import com.rtomyj.skc.model.ProductContent
import com.rtomyj.skc.skcsuggestionengine.traffic.TrafficService
import com.rtomyj.skc.util.enumeration.TrafficResourceType
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.stereotype.Service

@Service
class ProductService @Autowired constructor(
	@Qualifier("product-jdbc") private val productDao: ProductDao,
	private val trafficService: TrafficService
) {
	fun getSingleProductUsingLocale(productId: String, locale: String, clientIP: String): Product {
		GlobalScope.launch {
			trafficService.submitTrafficData(TrafficResourceType.PRODUCT, productId, clientIP)
		}

		val product = productDao.getProductInfo(productId, locale)

		product.productContent.addAll(productDao.getProductContents(productId, locale))
		product.setLinks()

		MonsterAssociation
			.transformMonsterLinkRating(
				product
					.productContent
					.stream()
					.filter { it.card != null }
					.map { productContent: ProductContent -> productContent.card!! }
					.toList()
			)
		return product
	}
}